iocage/lib/Release.py

```diff
@@ -0,0 +1,363 @@
+        helpers.init_host(self, host)
+
+        self.name = name
+        self._hashes = None
+        self._dataset = None
+        self.dataset = dataset
+        self.check_hashes = check_hashes is True
+
+        self._assets = ["base"]
+        if self.host.distribution.name != "HardenedBSD":
+            self._assets.append("lib32")
+
+    @property
+    def dataset(self):
+        if self._dataset is None:
+            self._dataset = self.zfs.get_dataset(self.dataset_name)
+        return self._dataset
+
+    @dataset.setter
+    def dataset(self, value):
+        if isinstance(value, libzfs.ZFSDataset):
+            try:
+                value.mountpoint
+            except:
+                value.mount()
+
+            self._dataset = value
+            self._update_name_from_dataset()
+
+        else:
+            self._zfs = None
+
+    @property
+    def releases_folder(self):
+        return self.host.datasets.releases.mountpoint
+
+    @property
+    def base_dataset(self):
+        # base datasets are created from releases. required to start
+        # zfs-basejails
+        return self.zfs.get_dataset(self.base_dataset_name)
+
+    @property
+    def base_dataset_name(self):
+        return f"{self.host.datasets.base.name}/{self.name}/root"
+
+    @property
+    def download_directory(self):
+        return f"{self.releases_folder}/{self.name}"
+
+    @property
+    def root_dir(self):
+        try:
+            return self.dataset.mountpoint
+        except:
+            return f"{self.releases_folder}/{self.name}/root"
+
+    @property
+    def assets(self):
+        return self._assets
+
+    @assets.setter
+    def assets(self, value):
+        value = [value] if isinstance(value, str) else value
+        self._assets = map(
+            lambda x: x if not x.endswith(".txz") else x[:-4],
+            value
+        )
+
+    @property
+    def mirror_url(self):
+        try:
+            if self._mirror_url:
+                return self._mirror_url
+        except:
+            pass
+        return self.host.distribution.mirror_url
+
+    @mirror_url.setter
+    def mirror_url(self, value):
+        url = urlparse(value)
+        if url.scheme not in self._supported_url_schemes:
+            raise Exception(f"Invalid URL scheme '{url.scheme}'")
+        self._mirror_url = url.geturl()
+
+    @property
+    def remote_url(self):
+        return f"{self.mirror_url}/{self.name}"
+
+    @property
+    def available(self):
+        try:
+            request = urllib.request.Request(self.remote_url, method="HEAD")
+            resource = urllib.request.urlopen(request)
+            return resource.getcode() == 200
+        except:
+            return False
+
+    @property
+    def fetched(self):
+        return os.path.isdir(self.root_dir)
+
+    @property
+    def zfs_pool(self):
+        try:
+            return self.host.datasets.releases.pool
+        except:
+            raise
+            pass
+
+        try:
```

Comment:
At the function end there is a custom exception that leaves a better hint what was wrong than the original exception(s).